### DISCUSSION TITLE
Upload file

### DISCUSSION THREAD

**Discussion by Alv-sys** [directus/directus#7234] [Feature Requests]
I am really loving Directus! Just I would like to mention an idea related with the File Upload UX.

![image](https://user-images.githubusercontent.com/47535920/128445735-2698f99a-0ce1-4218-ba85-cd7c735464a1.png)

To choose "Select File From Library" it is necessary to click the three points symbol, so the user is being directed to take the action to Upload the file, resulting in duplicated content.
The other case is that if you don't remember whether you uploaded the image, you have to take some seconds and think if you check the file folder or try to find the button to open it.

Would be nice to have that in a way to prioritize the use of the already uploaded images, some ideas:

Open the file library  directly:
![image](https://user-images.githubusercontent.com/47535920/128447093-f6a6ddc7-f261-4b66-80b5-ca087a84909d.png)

Give the option to open the library in the same area:
![image](https://user-images.githubusercontent.com/47535920/128448169-3277a2e5-2a87-460c-81c3-d02a07fb1371.png)

Or maybe leave the option visually available not mixed with the upload area:
![image](https://user-images.githubusercontent.com/47535920/128447410-cb2adc4a-103d-44c0-94e4-f5d76d6ffa75.png)

Cheers!

**Comment by benhaynes** [comment]
Thank you for this feedback! I think the best way to forefront this functionality (while keeping it unopinionated) is to have the entire interface click state become a dropdown. That would mean that when you click anywhere on the interface you get a dropdown of available options, including:

* Choose from device
* Choose from File Library
* Upload from URL

Thoughts on that approach?

**Reply by Alv-sys** [reply]
Hey @benhaynes, Thank you for your answer.
The idea you mention is a good option for sure!
In general, as a developer, I prefer option 1 of the images (open the library directly). In my experience the users for laziness they just upload the same files, plus it is easier for them to select the file than upload it, so it is a win-win (Resources-Ease of use).

But that is just my experience, I understand the objective of Directus, and  keeping it unopinionated fits better

Cheers, and thank you for reading!

**Reply by zhuyifang** [reply]
When you have a lot of pictures, it is almost impossible to find the duplicates

**Reply by Alv-sys** [reply]
> When you have a lot of pictures, it is almost impossible to find the duplicates

It is useful in several cases, if you are creating something and delete it, and you start to create it again, you still are able to find the file,  at the first position in the library, and you can easily reuse it. If you are creating a new element with no specific purpose, you can just choose from the already uploaded files, etc.
Normally for the case you mention, it is easy to recognize when you are able to see the final URL, so miimage.jpg and miimage.jpg2 makes you recognize immediately. Anyways I wouldn't like to divert the focus in the discussion with that.

**Reply by benhaynes** [reply]
For me, it's less about staying unopinionated, and more that I think there are a large number of users who are uploading a new file and need to find it on their device. So putting all these options at the same level should help.

Also, I think the issue of duplicates could (should!) be resolved via [additional logic](https://github.com/directus/directus/discussions/5162), not by the user. Then, if you upload a duplicate file the App could ask if you want to use the existing file instead (cancelling the upload).

**Comment by zhuyifang** [comment]
Here, I still hope to support:
The upload path (relative path) of the file upload component can be set in collections.

For example, I have configured the upload root directory in env, then here, I can set the subpath
@benhaynes

**Comment by Alv-sys** [comment]
Hi again!!

Has been a while since I posted this.
I think the mention to find duplicates diverted the topic, so just to keep it simple, let's talk just about the image interface.

My current library, with just 3 pages of content, it has 5 times the same image.

I could have even more, but sometimes I remember I already uploaded that image, but the default behavior of Uploading an Image makes me tend to upload every time.

When we are developing I guess we barely notice, but when you are using Directus as user, it really turns into a problem.
 
Just keeping it unopinionated would be great as your suggestion @benhaynes 
Any light on the horizon for this?


Thank you!

**Reply by benhaynes** [reply]
I think one option that would help significantly is to generate a hash of the file on upload. Then that hash could be compared to others in the library to see if a duplicate file already exists. I believe we started to have this feature in the past, but removed it due to the processing of hashes for large files. @azrikahar @rijkvanzanten thoughts?

**Reply by rijkvanzanten** [reply]
That's certainly one way. We do have to think about what we want to happen when you upload a file that already exists in that case. Does it automatically reuse the same file? Does it throw an error, then prompting the user in-app what they want to do? I imagine there might be a (hidden) use case somewhere for uploading duplicates on purpose?

**Reply by azrikahar** [reply]
Maybe we can take a similar approach like Activity & Revision Tracking, where it could be configurable:

- automatically reuse file with same hash
- prompt on each file (but I can also see this may potentially be too frequent and detrimental to users)
- don't check / allow duplicate

For point number 2, perhaps we can alleviate that by making it a per file/image interface setting (also fits OP's requirement), with "allow duplicate" as default to make it an opt-in thing.

**Reply by Alv-sys** [reply]
It would be great to have detection for duplicates! with a selection for that when creating the interface, it'd be out the box feature.
However, currently, the default box to upload is lacking the option of "Select From Library" at first sight, leading the upload of a picture. 

Do you guys think is possible to add something like this?
![image](https://user-images.githubusercontent.com/47535920/162641705-22edc643-4476-4eb6-ac00-3c9bc1158bf5.png)

Or any other idea to make it unopinionated?

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
